README.md
# :closed_lock_with_key::closed_lock_with_key::closed_lock_with_key: Get the Certificate Chain :closed_lock_with_key::closed_lock_with_key::closed_lock_with_key:

Getting the certificate chain for a hostname should not be difficult!

Sometimes websites do not present the full chain (against RFC recommendations) which means it's hard to find the certificate chain and troubleshoot.

This script will attempt to use the breadcrumbs that are left by the certificate to build the chain and output it into files.

It's worked on publicly available websites (at least the ones I've tested). Happy to troubleshoot if an error pops up.


# Requirements
```bash
$ python3 -m pip install re cryptography argparse ssl
```

# :runner: How to run
To get the arguments available `python3 getCertChain.py --help`. Should present information like this:
```bash
$ python3 getCertChain.py -h
usage: getCertChain.py [-h] [--hostname HOSTNAME] [--removeCertificateFiles] [--getCAcertPEM] [--insecure]

Get Certificate Chain v0.07

optional arguments:
  -h, --help            show this help message and exit
  --hostname HOSTNAME   The hostname:port pair that the script should connect to. Defaults to www.google.com:443.
  --removeCertificateFiles
                        Remove the certificate files in current working directory (*.crt, *.pem).
  --getCAcertPEM        Get cacert.pem from curl.se website to help find Root CA.
  --insecure            Allow insecure connections to establish.
```

# :books: Examples
To get the certificate chain for www.google.com
```bash
$ python3 getCertChain --hostname www.google.com
$
```
It by default will not display any output (unless it struggles to find the Root CA).

To see the files it created (from where the script ran):
```bash
$ ls -l
total 28
-rw-r--r-- 1 gituser gituser  1911 Mar 12 16:38 0-gts-root-r1.crt
-rw-r--r-- 1 gituser gituser  1996 Mar 12 16:38 1-gts-ca-1c3.crt
-rw-r--r-- 1 gituser gituser  1631 Mar 12 16:38 2-www.google.com.crt
-rw-r--r-- 1 gituser gituser 13385 Mar 12 16:36 getCertChain.py
```
If you notice the prefix of each new file that was created (i.e. 0-, 1-, 2-) this is the order of the chain with 
* `0-` being the Root CA
* `1-` being the intermediate CA
* `2-` being the endpoint/website certificate

The rest of the file name is deduced from the commonName in the certificate.

On some occasions, the intermediate CA does not present an Authority Information Access field (AIA) with a URI on where to get the file from. 

To help build the chain, we need some `external` help - leveraging the `cacert.pem` file from `curl.se` website.
This is the url: :link: `https://curl.se/ca/cacert.pem`

To get the `cacert.pem` file as part of the connection from `--hostname` argument:
```bash
$ python3 getCertChain.py --hostname www.google.com --getCAcertPEM
```

You could also run it by itself:
```bash
$ python3 getCertChain.py --getCAcertPEM
```

# :dash::hole: Clean up all the files
```bash
$ python3 getCertChain.py --removeCertificateFiles
Removing file 1-gts-ca-1c3.crt
Removing file 2-www.google.com.crt
Removing file 0-gts-root-r1.crt
Removing file cacert.pem
```

# :warning: MitM proxies/services
Given the way MitM proxies and decryption services work, this tool is not intended to be used behind devices that perform SSL/TLS decryption as those devices strip the `Authority Information Access` (AIA), and alter the `Subject Key Identifier` (SKI) and `Authority Key Identifier` (AKI) fields.

# Known issues 🌩️
If a website's certificate chain has 4 or more tiers and one of the intermediate CA's do not have a AIA field - the script will only show the website certificate and the first issuing CA. 

Something like this:
* Root CA
  * Issuing CA 1
    * Issuing Sub CA 1
      * Website certificate (example.com)

Script will only output `Issuing Sub CA 1` and `website certificate`.


CHANGELOG.md
# 2023-11-05
## Version 0.07
* Making the code more modular.
    * Created a `CertificateManager` class that handles certificate related operations.
    * Created a `FileManager` class that handles file operations
    * Created a `ConnectionManager` class that handles connections to URI's.


# 2023-08-13
## Version 0.06
* Fixing python syntax and code.


# 2023-07-09
## Version 0.05
* Adding ability to allow insecure connections (`--insecure`). Useful for when a website isn't properly configured.


# 2023-04-26
## Version 0.05
* Fixing bug [no-commonname](https://github.com/TheScriptGuy/getCertificateChain/issues/9)


# 2023-03-20
## Version 0.04
* Fixing bug [MitM proxy/ssl decrypt](https://github.com/TheScriptGuy/getCertificateChain/issues/5)


## Version 0.03
* Enhancement request [adding Root Found output](https://github.com/TheScriptGuy/getCertificateChain/issues/2)


## Version 0.02
* Fixing bug [deprecation warning](https://github.com/TheScriptGuy/getCertificateChain/issues/1)


# 2023-03-12
## Version 0.01
* First release.


tests/test_cert.py
import os
import pytest
from get_certificate_chain import (
    SSLCertificateChainDownloader,
)


@pytest.fixture
def root_ca_cert():
    with open(
        os.path.join(os.path.dirname(__file__), "test_data/root_ca_cert.pem"), "r"
    ) as f:
        cert = f.read()
    return cert


@pytest.fixture
def server_cert():
    with open(
        os.path.join(os.path.dirname(__file__), "test_data/server_cert.pem"), "r"
    ) as f:
        cert = f.read()
    return cert


@pytest.fixture
def cert_data(root_ca_cert, server_cert):
    return {
        "ca_cert_text": root_ca_cert,
        "cert_text": server_cert,
    }


def test_check_domain():
    downloader = SSLCertificateChainDownloader()

    # Test without port
    downloader.domain = "www.google.com"
    domain = downloader.check_domain()
    assert domain == {"hostname": "www.google.com", "port": 443}

    # Test with port
    downloader.domain = "www.google.com:8443"
    domain = downloader.check_domain()
    assert domain == {"hostname": "www.google.com", "port": 8443}


def test_normalize_subject():
    downloader = SSLCertificateChainDownloader()

    subject = "/C=US/ST=California/L=Mountain View/O=Google LLC/CN=www.google.com"
    assert (
        downloader.normalize_subject(subject)
        == "C_US_ST_California_L_Mountain_View_O_Google_LLC_CN_www_google_com"
    )


tests/generate_cert.sh
#!/bin/bash

# Create a self-signed certificate with the SKI extension
openssl req -x509 -new -nodes -days 365 -config example.cnf -out example1_cert.pem -extensions v3_ca


get_certificate_chain/tests/test_cert.py
import os
import pytest
from main import (
    SSLCertificateChainDownloader,
)


@pytest.fixture
def root_ca_cert():
    with open(
        os.path.join(os.path.dirname(__file__), "test_data/root_ca_cert.pem"), "r"
    ) as f:
        cert = f.read()
    return cert


@pytest.fixture
def server_cert():
    with open(
        os.path.join(os.path.dirname(__file__), "test_data/server_cert.pem"), "r"
    ) as f:
        cert = f.read()
    return cert


@pytest.fixture
def cert_data(root_ca_cert, server_cert):
    return {
        "ca_cert_text": root_ca_cert,
        "cert_text": server_cert,
    }


def test_check_domain():
    downloader = SSLCertificateChainDownloader()

    # Test without port
    downloader.domain = "www.google.com"
    domain = downloader.check_domain()
    assert domain == {"hostname": "www.google.com", "port": 443}

    # Test with port
    downloader.domain = "www.google.com:8443"
    domain = downloader.check_domain()
    assert domain == {"hostname": "www.google.com", "port": 8443}


def test_normalize_subject():
    downloader = SSLCertificateChainDownloader()

    subject = "/C=US/ST=California/L=Mountain View/O=Google LLC/CN=www.google.com"
    assert (
        downloader.normalize_subject(subject)
        == "C_US_ST_California_L_Mountain_View_O_Google_LLC_CN_www_google_com"
    )


main.py
"""
Download SSL certificate chain and save as PEM files.

This script connects to a given website, downloads its SSL certificate,
and saves it as a PEM file. If the certificate has an Authority Information
Access (AIA) extension, the script will download each certificate in the chain
and save them as PEM files as well.

"""
# Standard library imports
import glob
import os
import logging
import re
import ssl
import socket
import sys
from typing import Any, Dict, List, Union
from urllib.request import urlopen

# Third-party library imports
import argparse
from cryptography import x509
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import serialization
from cryptography.x509.oid import ExtensionOID

VERSION = "0.1.2"
CERT_CHAIN = []

# Configure logging
logging.basicConfig(
    level=logging.WARNING, format="%(asctime)s [%(levelname)s] %(message)s"
)


# parse arguments
def parse_arguments():
    """
    Parse command line arguments.

    Returns:
        argparse.Namespace: Parsed arguments.
    """
    parser = argparse.ArgumentParser(
        description="Export security rules and associated Security Profile Groups to a CSV file."
    )
    parser.add_argument(
        "--rm-ca-files",
        dest="remove_ca_files",
        action="store_true",
        help="Remove the cert files in current directory (*.crt, *.pem).",
    )
    parser.add_argument(
        "--get-ca-cert-pem",
        dest="get_ca_cert_pem",
        action="store_true",
        help="Get cacert.pem from curl.se website to help find Root CA.",
    )
    parser.add_argument(
        "--domain",
        dest="domain",
        default="www.google.com",
        help="The hostname to connect to. (default: %(default)s)",
    )
    return parser.parse_args()


class SSLCertificateChainDownloader:
    def __init__(self):
        self.cert_chain = []

    def remove_cacert_pem(self):
        """
        Remove certificate files from the current directory.
        """
        for crt_file in glob.glob("*.crt"):
            os.remove(crt_file)
        for pem_file in glob.glob("*.pem"):
            os.remove(pem_file)

    def get_cacert_pem(self):
        """
        Download the cacert.pem file from the curl.se website.
        """
        cacert_pem_url = "https://curl.se/ca/cacert.pem"
        cacert_pem_file = "cacert.pem"
        logging.info("Downloading %s to %s", cacert_pem_url, cacert_pem_file)
        with urlopen(cacert_pem_url) as response, open(
            cacert_pem_file, "wb"
        ) as out_file:
            if response.getcode() != 200:
                logging.error(
                    "Error downloading %s. HTTP status code: %s",
                    cacert_pem_url,
                    response.getcode(),
                )
                sys.exit(1)
            data = response.read()
            out_file.write(data)
        logging.info("Downloaded %s to %s", cacert_pem_url, cacert_pem_file)

    def check_domain(self) -> Dict[str, Any]:
        """
        Check and parse the domain provided by the user.

        Args:
            domain (str): The domain provided by the user.

        Returns:
            Dict[str, Any]: A dictionary containing the hostname and port.
        """
        hostname, _, port = self.domain.partition(":")
        return {"hostname": hostname, "port": int(port) if port else 443}

    def get_certificate(self, hostname: str, port: int) -> x509.Certificate:
        """
        Connect to a server and retrieve the SSL certificate.

        Args:
            hostname (str): The hostname to connect to.
            port (int): The port to connect to.

        Returns:
            x509.Certificate: The SSL certificate of the server.
        """
        try:
            context = ssl.create_default_context()
            with socket.create_connection((hostname, port)) as sock:
                with context.wrap_socket(sock, server_hostname=hostname) as ssl_socket:
                    cert_pem = ssl.DER_cert_to_PEM_cert(ssl_socket.getpeercert(True))
                    cert = x509.load_pem_x509_certificate(
                        cert_pem.encode(), default_backend()
                    )
            return cert
        except ConnectionRefusedError:
            print(f"Connection refused to {hostname}:{port}")
            sys.exit(1)
        except ssl.SSLError as e:
            print(f"SSL error: {e}")
            sys.exit(1)
        except socket.timeout:
            print(f"Connection timed out to {hostname}:{port}")
            sys.exit(1)
        except socket.gaierror:
            print(f"Hostname could not be resolved: {hostname}")
            sys.exit(1)

    def normalize_subject(self, subject: str) -> str:
        """
        Normalize the subject of a certificate.

        Args:
            subject (str): The subject of the certificate.

        Returns:
            str: The normalized subject.
        """
        return "_".join(
            part.strip()
            .replace("=", "_")
            .replace(".", "_")
            .replace(" ", "_")
            .replace(",", "_")
            for part in subject.split("/")
            if part.strip()
        )

    def save_ssl_certificate(
        self,
        ssl_certificate: x509.Certificate,
        file_name: str,
    ) -> None:
        """
        Save an SSL certificate to a file.

        Args:
            ssl_certificate (x509.Certificate): The SSL certificate to save.
            file_name (str): The file name to save the SSL certificate as.
        """
        with open(file_name, "wb") as f:
            f.write(ssl_certificate.public_bytes(encoding=serialization.Encoding.PEM))

    def write_chain_to_file(self, certificate_chain: List[x509.Certificate]) -> None:
        """
        Write a certificate chain to files.

        Args:
            certificate_chain (List[x509.Certificate]): The certificate chain to write to files.
        """
        for counter, certificate_item in enumerate(certificate_chain):
            cert_subject = certificate_item.subject.rfc4514_string()
            normalized_subject = self.normalize_subject(cert_subject)
            ssl_certificate_filename = (
                f"{len(certificate_chain) - 1 - counter}-{normalized_subject}.crt"
            )
            self.save_ssl_certificate(certificate_item, ssl_certificate_filename)

    def return_cert_aia(self, ssl_certificate: x509.Certificate) -> x509.Extensions:
        """
        Get the Authority Information Access (AIA) extension from a certificate.

        Args:
            ssl_certificate (x509.Certificate): The SSL certificate.

        Returns:
            x509.Extensions: The AIA extension or None if not found.
        """
        try:
            aia = ssl_certificate.extensions.get_extension_for_oid(
                ExtensionOID.AUTHORITY_INFORMATION_ACCESS
            )
            return aia
        except x509.ExtensionNotFound:
            return None

    def get_certificate_from_uri(self, uri: str) -> x509.Certificate:
        """
        Retrieve a certificate from the given URI.

        Args:
            uri (str): The URI to get the certificate from.

        Returns:
            x509.Certificate: The certificate from the URI or None if there was an error.
        """
        try:
            with urlopen(uri) as response:
                if response.getcode() != 200:
                    return None
                aia_content = response.read()
                ssl_certificate = ssl.DER_cert_to_PEM_cert(aia_content)
                cert = x509.load_pem_x509_certificate(
                    ssl_certificate.encode("ascii"), default_backend()
                )
                return cert
        except Exception:
            return None

    def return_cert_aia_list(self, ssl_certificate: x509.Certificate) -> list:
        """
        Get the list of AIA URIs from a certificate.

        Args:
            ssl_certificate (x509.Certificate): The SSL certificate.

        Returns:
            list: A list of AIA URIs.
        """
        aia_uri_list = []

        for extension in ssl_certificate.extensions:
            cert_value = extension.value

            if isinstance(cert_value, x509.AuthorityInformationAccess):
                data_aia = [x for x in cert_value or []]
                for item in data_aia:
                    if item.access_method._name == "caIssuers":
                        aia_uri_list.append(item.access_location._value)

        return aia_uri_list

    def return_cert_aki(self, ssl_certificate):
        """
        Get the Authority Key Identifier (AKI) from a certificate.

        Args:
            ssl_certificate (x509.Certificate): The SSL certificate.

        Returns:
            x509.AuthorityKeyIdentifier: The AKI extension or None if not found.
        """
        try:
            cert_aki = ssl_certificate.extensions.get_extension_for_oid(
                ExtensionOID.AUTHORITY_KEY_IDENTIFIER
            )
        except x509.extensions.ExtensionNotFound:
            cert_aki = None
        return cert_aki

    def return_cert_ski(self, ssl_certificate):
        """
        Get the Subject Key Identifier (SKI) from a certificate.

        Args:
            ssl_certificate (x509.Certificate): The SSL certificate.

        Returns:
            x509.SubjectKeyIdentifier: The SKI extension.
        """
        cert_ski = ssl_certificate.extensions.get_extension_for_oid(
            ExtensionOID.SUBJECT_KEY_IDENTIFIER
        )

        return cert_ski

    def load_root_ca_cert_chain(
        self,
        filename: str = None,
        ca_cert_text: str = None,
    ) -> Dict[str, str]:
        """
        Load the root CA certificate chain from a file or text.

        Args:
            filename (str, optional): The file name containing the root CA certificates.
            ca_cert_text (str, optional): The text containing the root CA certificates.

        Returns:
            Dict[str, str]: A dictionary containing the root CA certificates.
        """
        if filename is None and ca_cert_text is None:
            raise ValueError("Either filename or ca_cert_text must be provided")

        ca_root_store = {}

        if filename:
            with open(filename, "r") as f_ca_cert:
                ca_cert_text = f_ca_cert.read()

        lines = ca_cert_text.splitlines()
        line_count = len(lines)
        index = 0

        while index < line_count:
            current_line = lines[index]

            if re.search(r"^-----BEGIN CERTIFICATE-----", current_line):
                root_ca_cert = ""
                index += 1
                while index < line_count and not re.search(
                    r"^-----END CERTIFICATE-----", lines[index]
                ):
                    root_ca_cert += lines[index] + "\n"
                    index += 1

                root_ca_cert += lines[index] + "\n"
                index += 1

                cert = x509.load_pem_x509_certificate(
                    root_ca_cert.encode(), default_backend()
                )
                root_ca_name = cert.subject.rfc4514_string()

                ca_root_store[root_ca_name] = root_ca_cert
            else:
                index += 1

        print(f"Number of Root CA's loaded: {len(ca_root_store)}")
        return ca_root_store

    def walk_the_chain(
        self,
        ssl_certificate: x509.Certificate,
        depth: int,
        max_depth: int = 4,
    ):
        if depth <= max_depth:
            cert_aki = self.return_cert_aki(ssl_certificate)
            cert_ski = self.return_cert_ski(ssl_certificate)

            cert_aki_value = (
                cert_aki._value.key_identifier if cert_aki is not None else None
            )
            cert_ski_value = cert_ski._value.digest
            logging.info(
                f"Depth: {depth} - AKI: {cert_aki_value} - SKI: {cert_ski_value}"
            )

            if cert_aki_value is not None:
                aia_uri_list = self.return_cert_aia_list(ssl_certificate)
                if aia_uri_list:
                    for item in aia_uri_list:
                        next_cert = self.get_certificate_from_uri(item)

                        if next_cert is not None:
                            self.cert_chain.append(next_cert)
                            self.walk_the_chain(next_cert, depth + 1, max_depth)
                        else:
                            logging.warning("Could not retrieve certificate.")
                            sys.exit(1)
                else:
                    logging.warning("Certificate didn't have AIA.")
                    ca_root_store = self.load_root_ca_cert_chain("cacert.pem")
                    root_ca_cn = None

                    for root_ca in ca_root_store:
                        try:
                            root_ca_certificate_pem = ca_root_store[root_ca]
                            root_ca_certificate = x509.load_pem_x509_certificate(
                                root_ca_certificate_pem.encode("ascii")
                            )
                            root_ca_ski = self.return_cert_ski(root_ca_certificate)
                            root_ca_ski_value = root_ca_ski._value.digest

                            if root_ca_ski_value == cert_aki_value:
                                root_ca_cn = root_ca
                                self.cert_chain.append(root_ca_certificate)
                                logging.info(
                                    f"Root CA Found - {root_ca_cn}\nCERT_CHAIN - {self.cert_chain}"
                                )
                                break
                        except x509.extensions.ExtensionNotFound:
                            logging.info("Root CA didn't have a SKI. Skipping...")
                            pass

                    if root_ca_cn is None:
                        logging.error("Root CA NOT found.")
                        sys.exit(1)

    def run(self, args: Union[argparse.Namespace, dict]):
        if isinstance(args, argparse.Namespace):
            self.domain = args.domain
        elif isinstance(args, dict):
            self.domain = args.get("domain")
        else:
            raise ValueError(
                "Invalid argument type. Expected argparse.Namespace or dict."
            )

        self.parsed_domain = self.check_domain()

        if isinstance(args, argparse.Namespace):
            remove_ca_files = args.remove_ca_files
            get_ca_cert_pem = args.get_ca_cert_pem
        else:
            remove_ca_files = args.get("remove_ca_files")
            get_ca_cert_pem = args.get("get_ca_cert_pem")

        if remove_ca_files:
            self.remove_ca_files()

        if get_ca_cert_pem:
            self.get_ca_cert_pem()

        ssl_certificate = self.get_certificate(
            self.parsed_domain["hostname"], self.parsed_domain["port"]
        )

        aia = self.return_cert_aia(ssl_certificate)

        if aia is not None and not self.return_cert_aia(ssl_certificate):
            logging.error(
                "Could not find AIA, possible decryption taking place upstream?"
            )
            sys.exit(1)

        self.cert_chain.append(ssl_certificate)

        self.walk_the_chain(ssl_certificate, 1, max_depth=4)

        self.write_chain_to_file(self.cert_chain)

        logging.info("Certificate chain downloaded and saved.")


def main() -> None:
    """
    Main function to execute the script. Parses arguments, retrieves the SSL certificate, walks the chain,
    and writes the certificate chain and PEM-encoded certificates.
    """
    args = parse_arguments()
    downloader = SSLCertificateChainDownloader()
    downloader.run(args)


if __name__ == "__main__":
    main()


FileManager.py
import os
import glob
from typing import List, Any


class FileManager:
    """
    FileManager class that provides methods for deleting files and writing objects to files.
    """

    version = '1.0'

    @staticmethod
    def delete_files(files_or_wildcards: List[str]) -> None:
        """
        Deletes files based on a list of filenames or wildcards. If a file does not exist,
        or the wildcard does not yield any results, it prints a message with the filename or wildcard.

        Parameters:
            files_or_wildcards (List[str]): A list of filenames or wildcard strings to match files for deletion.

        Returns:
            None
        """
        for pattern in files_or_wildcards:
            # Expand wildcard to actual file list
            files = glob.glob(pattern)
            if not files:
                print(f"No files found for wildcard/pattern: '{pattern}'")
                continue

            for file in files:
                try:
                    os.remove(file)
                    print(f"File deleted: {file}")
                except FileNotFoundError:
                    print(f"File does not exist and cannot be deleted: {file}")
                except OSError as e:
                    print(f"Error deleting file {file}: {e}")

    @staticmethod
    def write_to_file(obj: Any, filename: str) -> None:
        """
        Writes an object to a file with the specified filename. If the object is a string or bytes,
        it writes directly to the file. If it's another type that can be converted to a string (e.g., dict, list),
        it converts to a string using str() before writing.

        Parameters:
            obj (Any): The object to be written to the file. Can be a string, bytes, or any object
                       that has a string representation.
            filename (str): The name of the file to which the object will be written.

        Returns:
            None
        """
        try:
            # Check if the object is binary or text and open the file in the appropriate mode
            mode = 'wb' if isinstance(obj, bytes) else 'w'
            with open(filename, mode) as file:
                # If the object is not bytes, convert it to string
                if not isinstance(obj, bytes):
                    obj = str(obj)
                file.write(obj)
                print(f"Object successfully written to {filename}")
        except IOError as e:
            print(f"Failed to write to file {filename}: {e}")
        except TypeError as e:
            print(f"TypeError: The object could not be written to the file {filename}: {e}")


ConnectionManager.py
import ssl
import socket
import requests
import sys

from cryptography import x509
from typing import Optional


class ConnectionManager:
    """ConnectionManager Class"""

    version = "0.1"

    def __init__(self) -> None:
        """Initialize the ConnectionManager class."""
        self.starting_hostname = ""

    def configure_hostname(self, __hostname) -> None:
        """Configure the hostname."""
        self.starting_hostname = self.reformat_hostname(__hostname)

    @staticmethod
    def reformat_hostname(__hostname) -> dict:
        """
        Parse __hostname argument.
        Make sure that if a port isn't supplied, then assume default port of 443.
        Return the hostname:port combination as a dict type.
        """
        tmpLine = ""
        hostnameQuery = ""

        # If the ':' is in the hostname argument, then we'll assume it's meant to be a port following the ':'.
        if ":" in __hostname:
            tmpLine = __hostname.split(':')
            hostnameQuery = {"hostname": tmpLine[0], "port": int(tmpLine[1])}
        else:
            # If no ':' is found, then set default port 443.
            hostnameQuery = {"hostname": __hostname, "port": 443}

        return hostnameQuery

    def get_certificate(self, secure: bool = True) -> x509.Certificate:
        """Retrieves the certificate from a hostname:port pair."""
        sslCertificate = None
        try:
            # Create the SSL context
            if secure:
                sslContext = ssl.create_default_context()
            else:
                sslContext = ssl._create_unverified_context()

            with socket.create_connection((self.starting_hostname['hostname'], self.starting_hostname['port'])) as sock:
                with sslContext.wrap_socket(sock, server_hostname=self.starting_hostname['hostname']) as sslSocket:
                    # Get the certificate from the connection, convert it to PEM format.
                    sslCertificate = ssl.DER_cert_to_PEM_cert(sslSocket.getpeercert(True))

            # Load the PEM formatted file.
            sslCertificate = x509.load_pem_x509_certificate(sslCertificate.encode('ascii'))

        except ssl.SSLCertVerificationError as e:
            print(f"SSL Verification error. {e.verify_message}\nTry with the --insecure option.")
            sys.exit(1)
        except ConnectionRefusedError:
            print(f"Connection refused to {self.starting_hostname['hostname']}:{__port}")
            sys.exit(1)

        # Return the sslCertificate object.
        return sslCertificate

    @staticmethod
    def get_file_contents(uri: str) -> Optional[bytes]:
        """
        Retrieve the file contents from a specified URI if the HTTP response code is 200.

        Parameters:
            uri (str): The Uniform Resource Identifier (URI) from which to retrieve the file contents.

        Returns:
            Optional[bytes]: The file contents as bytes if retrieval is successful; None otherwise.

        Raises:
            ValueError: If the URI is empty or None.
            HTTPError: If the HTTP response code is not 200.
            RequestException: If there's an issue with the network request.
        """
        if not uri:
            raise ValueError("The URI must not be empty.")

        try:
            response = requests.get(uri)
            if response.status_code != 200:
                # If the status is not 200, you can raise an HTTPError
                response.raise_for_status()
        except HTTPError as e:
            # Handle specific HTTP errors if needed
            print(f"Received non-200 HTTP status code: {e}")
            sys.exit(1)
        except RequestException as e:
            # Handle any other requests-related exceptions
            print(f"An error occurred while trying to retrieve the file: {e}")
            sys.exit(1)

        return response.content

    def get_certificate_from_uri(self, __uri: str) -> x509.Certificate:
        """Gets the certificate from a URI.
        By default, we're expecting to find nothing. Therefore certI = None.
        If we find something, we'll update certI accordingly.
        """
        certI = None

        # Attempt to get the aia from __uri
        aiaRequest = requests.get(__uri)

        # If response status code is 200
        if aiaRequest.status_code == 200:
            # Get the content and assign to aiaContent
            aiaContent = aiaRequest.content

            # Convert the certificate into PEM format.
            sslCertificate = ssl.DER_cert_to_PEM_cert(aiaContent)

            # Load the PEM formatted content using x509 module.
            certI = x509.load_pem_x509_certificate(sslCertificate.encode('ascii'))

        # Return certI back to the script.
        return certI


getCertChain.py
# Description:     Get the certificate chain from a website.
# Author:          TheScriptGuy
# Last modified:   2023-11-04
# Version:         0.07

import argparse
import sys
import os

from ConnectionManager import ConnectionManager
from FileManager import FileManager
from CertificateManager import CertificateManager

scriptVersion = "0.07"
maxDepth = 4
certChain = []


def parseArguments():
    """Create argument options and parse through them to determine what to do with script."""
    # Instantiate the parser
    parser = argparse.ArgumentParser(description=f'Get Certificate Chain v{scriptVersion}')

    # Optional arguments
    parser.add_argument('--hostname', default='www.google.com:443',
                        help='The hostname:port pair that the script should connect to. Defaults to www.google.com:443.')

    parser.add_argument('--removeCertificateFiles', action='store_true',
                        help='Remove the certificate files in current working directory (*.crt, *.pem).')

    parser.add_argument('--getCAcertPEM', action='store_true',
                        help='Get cacert.pem from curl.se website to help find Root CA.')

    parser.add_argument('--insecure', action='store_true',
                        help='Allow insecure connections to establish.')

    global args

    args = parser.parse_args()


def getCAcertPEM() -> None:
    """Retrieves the cacert.pem file from curl.se website."""
    connection_manager = ConnectionManager()
    file_manager = FileManager()

    file_manager.write_to_file(
        connection_manager.get_file_contents("https://curl.se/ca/cacert.pem"),
        "cacert.pem"
        )


def main():
    """Main subroutine."""
    # Parse the arguments
    parseArguments()

    # If --removeCertificateFiles is passed, remove files and exit
    if args.removeCertificateFiles:
        file_manager = FileManager()
        file_manager.delete_files(["*.crt", "*.pem"])
        sys.exit(0)

    if args.getCAcertPEM:
        getCAcertPEM()

    # Check to see if hostname argument is passed. If not, exit.
    if args.hostname:
        # Define a connection_manager object as ConnectionManager class
        connection_manager = ConnectionManager()
        connection_manager.configure_hostname(args.hostname)
    else:
        # No hostname argument was supplied. Exit.
        print("Please supply a --hostname argument.")
        sys.exit(1)

    # Define a certificate_manager object from CertificateManager class
    certificate_manager = CertificateManager("cacert.pem", maxDepth)

    # Get the website certificate object, by default it's a secure connection.
    if args.insecure:
        certificate = connection_manager.get_certificate(secure=False)
    else:
        certificate = connection_manager.get_certificate()

    certificate_manager.set_certificate(certificate)

    # Prepare to walk the chain
    certificate_manager.start_walk_of_chain()

    # Now lets write the contents to a file
    certificate_manager.write_chain_to_file()


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('Interrupted')
        print()
        try:
            sys.exit(0)
        except SystemExit:
            os.exit(0)
